src/main/java/com/minecolonies/structures/helpers/Structure.java

```diff
@@ -82,63 +92,361 @@
+    }
+
+    /**
+     * get a input stream for a schematic within a specif folder.
+     *
+     * @param folder        where to load it from.
+     * @param structureName name of the structure to load.
+     * @return the input stream or null
+     */
+    private static InputStream getStreamFromFolder(final File folder, final String structureName)
+    {
+        try
+        {
+            if (folder.exists())
+            {
+                //We need to check that we stay within the correct folder
+                final File nbtFile = new File(folder.getPath() + "/" + structureName + ".nbt");
+                if (nbtFile.toURI().normalize().getPath().startsWith(folder.toURI().normalize().getPath()))
                 {
-                    inputstream = new FileInputStream(decorationFolder.getPath() + "/" + structureName + ".nbt");
+                    return new FileInputStream(folder.getPath() + "/" + structureName + ".nbt");
                 }
                 else
                 {
-                    throw new FileNotFoundException("Unable to find structure: " + structureName);
+                    Log.getLogger().error("Structure: Illegal structure name \"" + structureName + "\"");
                 }
             }
-            catch (final FileNotFoundException e)
+            else
             {
-                Log.getLogger().warn("Couldn't find any structure with this name anywhere", e);
+                throw new FileNotFoundException("Unable to find structure: " + structureName);
             }
         }
+        catch (final FileNotFoundException e)
+        {
+            //Ignore the error
+        }
+        return null;
+    }
 
-        if (inputstream == null)
+    /**
+     * get a input stream for a schematic from jar.
+     *
+     * @param structureName name of the structure to load from the jar.
+     * @return the input stream or null
+     */
+    private static InputStream getStreamFromJar(final String structureName)
+    {
+        return MinecraftServer.class.getResourceAsStream("/assets/" + Constants.MOD_ID + '/' + structureName + ".nbt");
+    }
+
+    /**
+     * get the Template from the structure.
+     *
+     * @return The templae for the structure
+     */
+    public Template getTemplate()
+    {
+        return this.template;
+    }
+
+    /**
+     * Convert an InputStream into and array of bytes.
+     *
+     * @param stream to be converted to bytes array
+     * @return the array of bytes, array is size 0 when the stream is null
+     */
+    public static byte[] getStreamAsByteArray(final InputStream stream)
+    {
+        if (stream == null)
         {
-            return;
+            Log.getLogger().info("Structure.getStreamAsByteArray: stream is null this should not happen");
+            return new byte[0];
+        }
+        try
+        {
+            ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+
+            int nRead;
+            byte[] data = new byte[16384];
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Assign this magic number 16384 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Add underscores to this numeric value for readability [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2148)